Add spec generator for migration generator

NickLaMuro opened the pull request:
Overwrites `rails generate migration` to allow a spec file to be generated when a migration is generated with the same file name.

QA
--

To test, you can run something like the following:

```ruby
$ bin/rails g migration foo_bar_baz -p
      create  db/migrate/20210727212808_foo_bar_baz.rb
      create  spec/migrations/20210727212808_foo_bar_baz_spec.rb
```

And just validate that the file names are as you would expect.  To validate the files are created properly, just remove the `-p` argument:

```ruby
$ bin/rails g migration test_two
      create  db/migrate/20210727213008_test_two.rb
      create  spec/migrations/20210727213008_test_two_spec.rb
```

Closes https://github.com/ManageIQ/manageiq-schema/issues/479

Fryguy commented on lib/generators/migration/migration_generator.rb:
```diff
@@ -0,0 +1,42 @@
+require 'rails/generators/active_record/migration/migration_generator'
+
+# Overwirtes the default `ActiveRecord::Generators::MigrationGenerator` to
```
```suggestion
# Overwrites the default `ActiveRecord::Generators::MigrationGenerator` to
```

:trollface:

Fryguy commented on lib/generators/migration/templates/migration_spec.rb.tt:
```diff
@@ -0,0 +1,29 @@
+  # let(:my_model_stub) { migration_stub(:MyModel) }
+
+  migration_context :up do
+    # Create data
```
```suggestion
    # Create data
    #
    # Example:
    #
    #   my_model_stub.create!(attributes)
```

NickLaMuro replied:
👍

I think I stopped filling this out as it took me forever to figure out how to actually override it properly.

Fryguy commented:
This is pretty cool.  We may also want a note inside the generator file that the spec is only necessary for data migrations and should not be included for "normal" Rails migrations.

NickLaMuro commented on spec/lib/generators/migration_specs_generator_spec.rb:
```diff
@@ -0,0 +1,32 @@
+  it "runs the generator without errors" do
+    expect {
+      run_generator
+    }.not_to raise_error
```
Not entirely sure these specs even would run (if I didn't have the old model names here).

Will see if they do, but if not, I will probably delete this file.

NickLaMuro replied:
Okay, got this working now.  Resolving.

kbrock commented:
Thank you.
You have made life for the developers just a little bit better. one migration at a time.

kbrock commented on lib/generators/migration/templates/migration_spec.rb.tt:
```diff
@@ -0,0 +1,37 @@
+require_migration
```
I'm still very proud of this line -- so much magic to get us here

Fryguy commented on lib/generators/migration/migration_generator.rb:
```diff
@@ -0,0 +1,56 @@
+    end
+
+    def self.miq_schema_root
+      File.expand_path(File.join("..", "..", ".."), __dir__)
```
Would this be simpler?
```suggestion
      ManageIQ::Schema::Engine.root
```

Then you also have a Pathname object, so the other calls like db_migrate_path can piggyback on that

```ruby
def db_migrate_path
  miq_schema_root.join("db", "migrate")
end
```

NickLaMuro replied:
Done.

Fryguy commented on lib/generators/migration/templates/migration_spec.rb.tt:
```diff
@@ -0,0 +1,37 @@
+require_migration
+
+describe <%= spec_class_name %> do
+  # let(:<%= spec_symbol_name %>_stub) { migration_stub(:<%= spec_class_name %>) }
```
I don't think this is right - it looks like the stub has to do with the migration name, but it is kind of unrelated - it's more about the models defined in the migration.  I think the old way with the hardcode MyModelStub looked clearer.

Fryguy commented on spec/lib/generators/migration_generator_spec.rb:
```diff
@@ -0,0 +1,57 @@
+  
+  before do 
+    prepare_destination
+    allow(described_class).to receive(:miq_schema_root).and_return(destination_root)
```
_Technically_ this is returning a Pathname where the original method didn't, which makes them unmatched.  However, if you change it as I suggested in https://github.com/ManageIQ/manageiq-schema/pull/592/files#r678321714, then they'd both be Pathnames.

NickLaMuro replied:
Done.

Fryguy commented on lib/generators/migration/migration_generator.rb:
```diff
@@ -0,0 +1,55 @@
+
+    def self.miq_schema_root
+      ManageIQ::Schema::Engine.root
+      # File.expand_path(File.join("..", "..", ".."), __dir__)
```
I think these commented bits can go away now.

NickLaMuro replied:
F... I kept these as comments there in case using `Pathname` was going to be a problem, but forgot to remove them once the tests passed.

Thanks for the ping.  Resolving.

Fryguy commented:
Updated the OP to link https://github.com/ManageIQ/manageiq-schema/issues/479

Fryguy commented:
@NickLaMuro Can you explain the `-p` argument in the OP?  I feel like _not_ having it as the default will cause more trouble, because it creates an invalid spec test.

NickLaMuro commented:
@Fryguy it is short for `--pretend`, which means it is a `--no-op`.

Fryguy commented:
I think you must have a typo in the OP then?

```
$ bin/rails g migration test_two
      create  db/migrate/20210727213008_test_two.rb
      create  spec/migrations/20210727213008_test_two.rb
```

That second one should end in _spec.rb, right?

NickLaMuro commented:
@Fryguy Oh right, you are correct. Fixed

(and yes, I just took the first one and replaced output without actually running it... the secret is out...)

jrafanie commented:
This is sweet,  nice work @NickLaMuro
